src/core/execute.c

```diff
@@ -5145,6 +5226,24 @@ static int exec_child(
                                 LOG_UNIT_MESSAGE(unit, "Executing: %s", line));
         }
 
+        /* If we are a user manager and setting up a UID/GID full mapping from a user namespace PID1 created for us,
+         * after we are done setting up we want to go back to the original UID/GID, given we'll be root when we
+         * move into the namespace. The point of having PID1 set us a mapping for us is to make sure there's no
+         * visible difference for the units due to the implicit user namespace. */
+        if (!MANAGER_IS_SYSTEM(unit->manager) && userns_set_up && !context->private_users) {
+                r = setgid(saved_gid);
+                if (r < 0) {
+                        *exit_status = EXIT_USER;
+                        return log_unit_error_errno(unit, errno, "Failed to set GID back to " GID_FMT ": %m", saved_gid);
+                }
+
+                r = setuid(saved_uid);
+                if (r < 0) {
+                        *exit_status = EXIT_USER;
+                        return log_unit_error_errno(unit, errno, "Failed to set UID back to " UID_FMT ": %m", saved_uid);
+                }
```

Comment:
hmm, i don't follow here. in the classic code paths we should have done this too, no?

i..e in the status quo ante if we use PrivateUsers= we currently run as root inside the userns? we really shouldn, we should be the service user there too...

Reply:
In the status quo it depends on the `User=` property, as expected, because the userns comes later

Reply:
hmm, that is a bit icky no? we probably should ensure that if we open a userns we *always* do at least one setresuid() afterwards.